### PR TITLE
JS: support sanitizers that sanitize individual chars in js/shell-commandconstructed-from-input

### PR DESCRIPTION
**Pull request by erik-krogh**
Gets a TN for CVE-2020-26274

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/security/dataflow/UnsafeShellCommandConstructionCustomizations.qll [hunk 1]
```diff
@@ -191,4 +194,37 @@ module UnsafeShellCommandConstruction {
...
   }
+
+  private import semmle.javascript.dataflow.internal.AccessPaths
+  private import semmle.javascript.dataflow.InferredTypes
+
+  /**
+   * Holds if `instance` is an instance of the access-path `ap`, and there exists a guard
+   * that ensures that `instance` is not equal to `char`.
+   */
+  private predicate blocksCharInAccess(AccessPath ap, string char, Expr instance) {
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/security/dataflow/UnsafeShellCommandConstructionCustomizations.qll:205]: If the performance works out, we should try to move this `SanitizedChar` and `blocksCharInAccess` to `TaintTracking.qll` (lets make a separate issue for that).

File: javascript/ql/src/semmle/javascript/security/dataflow/UnsafeShellCommandConstructionCustomizations.qll [hunk 2]
```diff
@@ -191,4 +194,37 @@ module UnsafeShellCommandConstruction {
       )
     }
   }
+
+  private import semmle.javascript.dataflow.internal.AccessPaths
+  private import semmle.javascript.dataflow.InferredTypes
+
+  /**
+   * Holds if `instance` is an instance of the access-path `ap`, and there exists a guard
+   * that ensures that `instance` is not equal to `char`.
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/security/dataflow/UnsafeShellCommandConstructionCustomizations.qll:203]: I could use a source code example example here

File: javascript/ql/src/semmle/javascript/security/dataflow/UnsafeShellCommandConstructionCustomizations.qll [hunk 3]
```diff
@@ -191,4 +194,37 @@ module UnsafeShellCommandConstruction {
...
+
+  /**
+   * A sanitizer for a single char, where the char cannot be an unsafe shell char.
+   */
+  class SanitizedChar extends Sanitizer, DataFlow::ValueNode {
+    override PropAccess astNode;
+
+    SanitizedChar() {
+      exists(AccessPath ap | this.asExpr() = ap.getAnInstance() |
+        forall(string char | char = getAShellChar() | blocksCharInAccess(ap, char, astNode))
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/security/dataflow/UnsafeShellCommandConstructionCustomizations.qll:225]: Drive-by: have you seen any evidence that a shell variant of `js/incomplete-url-scheme-check` could be interesting?

**Review by esbena — COMMENTED** [review]
LGTM except for a nit, two ideas for future work.

File: javascript/ql/src/semmle/javascript/security/dataflow/UnsafeShellCommandConstructionCustomizations.qll (hunk 3 above)
Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/security/dataflow/UnsafeShellCommandConstructionCustomizations.qll:225]: Not enough that I think we should make a query for it.   

But the [fix commit to `CVE-2020-26274`](https://github.com/sebhildebrandt/systeminformation/commit/1faadcbf68f1b1fdd5eb2054f68fc932be32ac99) was such an incomplete sanitizer.

Comment (esbena) [reply javascript/ql/src/semmle/javascript/security/dataflow/UnsafeShellCommandConstructionCustomizations.qll:225]: I have created an issue for it, but not assigned anyone.

File: javascript/ql/src/semmle/javascript/security/dataflow/UnsafeShellCommandConstructionCustomizations.qll [hunk 4]
```diff
@@ -208,4 +211,50 @@ module UnsafeShellCommandConstruction {
...
+      guard.getTest() = test and
+      guard.dominates(bb) and
+      test.getAnOperand() = ap.getAnInstance() and
+      instance = ap.getAnInstanceIn(bb) and
+      guard.getOutcome() != test.getPolarity()
+    )
+  }
+
+  /**
+   * A sanitizer for a single char, where the char cannot be an unsafe shell char.
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/security/dataflow/UnsafeShellCommandConstructionCustomizations.qll:248]:
```suggestion
   * A sanitizer for a single character, where the character cannot be an unsafe shell character.
```
Sorry. Should have said "ditto" below above.

**Review by esbena — APPROVED** [review]

**Comment by erik-krogh** [comment]
[Evaluation looks fine](https://git.semmle.com/erik/dist-compare-reports/tree/eriks-profile-machine_1610060169513).

File: javascript/ql/src/semmle/javascript/security/dataflow/UnsafeShellCommandConstructionCustomizations.qll (hunk 1 above)
Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/security/dataflow/UnsafeShellCommandConstructionCustomizations.qll:235]: So generalize the sanitizer such that other queries can use it?

**Pull request merged**